### PR TITLE
Set native flags according to CMAKE_SYSTEM_PROCESSOR

### PR DESCRIPTION
**Pull request by masterleinad**
Alternative to #5809, fixing #5808, following https://github.com/kokkos/kokkos/pull/5809#issuecomment-1410832647.
For x86 targets, we use `-march=native -mtune=native` and otherwise `-mcpu=native`.

### CODE REVIEW

File: cmake/kokkos_functions.cmake [hunk 1]
```diff
@@ -827,7 +827,7 @@ FUNCTION(kokkos_link_tpl TARGET)
 ENDFUNCTION()
 
 FUNCTION(COMPILER_SPECIFIC_OPTIONS_HELPER)
-  SET(COMPILERS NVIDIA NVHPC DEFAULT Cray Intel Clang AppleClang IntelLLVM GNU HIPCC Fujitsu MSVC)
+  SET(COMPILERS NVIDIA NVHPC DEFAULT Cray Intel Clang AppleClang IntelLLVM GNU HIPCC Fujitsu MSVC ARMClang)
```
Comment (dalg24) [inline cmake/kokkos_functions.cmake:830]: No point on adding
```suggestion
  SET(COMPILERS NVIDIA NVHPC DEFAULT Cray Intel Clang AppleClang IntelLLVM GNU HIPCC Fujitsu MSVC)
```

File: cmake/kokkos_arch.cmake [hunk 2]
```diff
@@ -229,9 +229,14 @@ IF(KOKKOS_ARCH_NATIVE)
     MESSAGE(FATAL_ERROR "MSVC doesn't support ARCH_NATIVE!")
   ENDIF()
 
+  IF(CMAKE_SYSTEM_PROCESSOR MATCHES "(x86)|(X86)|(amd64)|(AMD64)")
```
Comment (dalg24) [inline cmake/kokkos_arch.cmake:232]: Consider making the string all uppercase before to avoid vexing `(Amd64)` not matching

**Comment by masterleinad** [comment]
Only `CUDA-11.7-NVHPC` is failing in
```
- KokkosCore_UnitTest_CudaTimingBased (Failed)
```

**Comment by masterleinad** [comment]
@antoine-morvan Would you mind trying this and report back?

**Comment by antoine-morvan** [comment]
Works fine on x86_64 and aarch64. Tried:
* x86_64: GCC 12; LLVM 14 & 15; OneAPI 2022.3 (icpc & icpx); NVHPC 22.9; AOCC 4.0
* aarch64: GCC 12; LLVM 14 & 15; ARMClang 22.1

Failing with NVHPC 22.9 on aarch64, as `-mcpu/march/mtune` are alias for `-tp` ... but not present on all releases ...
This commit would most likely fix the issue: https://github.com/kokkos/kokkos/commit/3c5e0530d82f807c338b674e51ec704a989b6397

**Review by dalg24 — APPROVED** [review]
Lets merge as is and handle NVPHC in a follow-up PR

**Review by crtrott — APPROVED** [review]

**Pull request merged**